### PR TITLE
virt-handler: Restrict node access

### PR DESCRIPTION
**Pull request by RamLavi**
### What this PR does
As of today a virt-handler has "full" access to all Nodes. Therefore installing Kubevirt means that every node will have a service account that can be miss used to influence the scheduling of other Pods, potentially with a more privileged service account.

This PR introduces [validationAdmissionPolicy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/), that will prevent the virt-handler user from:
- changing the node's spec
- adding/removing/updating labels and annotations not owned by kubevirt.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:
- This change will take affect only on cluster that support validatingAdmissionPolicy v1. 
- e2e test were introduced, but due to bazel related [issue](https://github.com/google/cel-go/issues/947) - unit tests will be introduced in later PR (see draft [here](https://github.com/kubevirt/kubevirt/pull/12021))

The following alternatives were considered: Full alternative discussion on [proposal PR](https://github.com/kubevirt/community/pull/246)

This PR is relying on https://github.com/kubevirt/kubevirt/pull/11840, ergo two first commits are not part of this PR.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Introduce validatingAdmissionPolicy to restrict node patches on virt-handler
```

### CODE REVIEW

**Comment by RamLavi** [comment]
/removing WIP
Adding hold, until https://github.com/kubevirt/kubevirt/pull/11840 is merged

/hold

**Comment by RamLavi** [comment]
/test pull-kubevirt-e2e-k8s-1.30-sig-compute
checking that the test runs at all

File: tests/validatingaddmisionpolicy/noderestrictions.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"kubevirt.io/kubevirt/pkg/virt-operator/resource/generate/components"
+	"kubevirt.io/kubevirt/pkg/virt-operator/util"
+	"kubevirt.io/kubevirt/tests/exec"
+	"kubevirt.io/kubevirt/tests/framework/kubevirt"
+	"kubevirt.io/kubevirt/tests/libnode"
+)
+
+var _ = Describe("[sig-compute] virt-handler node restrictions via validatingAdmissionPolicy", func() {
```
Comment (xpivarc) [inline tests/validatingaddmisionpolicy/noderestrictions.go:39]: `decorators.SigCompute` is missing.

**Comment by RamLavi** [comment]
should skip on:
/test pull-kubevirt-e2e-k8s-1.29-sig-compute

should pass on:
/test pull-kubevirt-e2e-k8s-1.30-sig-compute

**Comment by fossedihelm** [comment]
/unhold
@xpivarc PTAL

**Review by xpivarc — COMMENTED** [review]
/approve

**Review by jean-edouard — APPROVED** [review]

**Comment by RamLavi** [comment]
fail is unrelated
/test pull-kubevirt-e2e-k8s-1.30-sig-compute-serial

**Comment by fabiand** [comment]
@RamLavi how does an admin know that this feature is in use or not?

How does an admin know - if VAP is not enabled - that KubeVirt has more security to offer if VAPs were enabled?

**Comment by RamLavi** [comment]
> @RamLavi how does an admin know that this feature is in use or not?
> 
> How does an admin know - if VAP is not enabled - that KubeVirt has more security to offer if VAPs were enabled?

VAPv1 is enabled (GA) on k8s1.30+, so it is a no-op for the admin. 
The only issue is with the kubevirt U/S constraint to support 2 releases backwards. As discussed with @acardace - the bug fix errata should explicitly mention that the fix will take affect only on clusters of k8s1.30+.

**Comment by fabiand** [comment]
@RamLavi but how does an admin know that this feature is enabled? How do I know for sure?
I guess VAP can also be disabled by a Kube admin, how does the admin then know that this security issue exists?

Also: What happens if kubevirt is run on kube without VAP present/enabled?

**Comment by RamLavi** [comment]
> @RamLavi but how does an admin know that this feature is enabled? 

@fabiand - please note that this is not a feature, it's a bug fix, a security bug fix.

> How do I know for sure? 

If the k8s installed is 1.30+, then VAPv1 k8s resource is available by default, and the admin can assume it is fixed. 
The admin can also get the deployed VAPv1 resource we deploy and see that it is indeed deployed.

> I guess VAP can also be disabled by a Kube admin, how does the admin then know that this security issue exists?

This is kind of an odd question. I don't see a reason why the admin would disable a GAed k8s feature like VAPv1 (or if it is even possible to disable it once it is GA). 
The important thing is that the fix is **not** something that the admin needs to manually "register" to (cause then 99% of the admins won't notice it). The approach Kubevirt is taking in this PR is: if the resource is available - then it would deploy it. 

> 
> Also: What happens if kubevirt is run on kube without VAP present/enabled?

then the VAP resource will not be deployed. A similar approach is taken with prometheus rules: if prometheus CRD is not deployed on the cluster - then kubevirt will not deploy the rules CR. The big difference is that VAPv1 is a k8s resource (not a CRD)

**Pull request merged**